dev/sg/sg_check.go

```diff
@@ -0,0 +1,197 @@
+				FlagSet:   checkGoFlagSet,
+				Exec:      runCheckScriptsAndReport(scriptChecks["go"]...),
+			},
+			{
+				Name:      "docsite",
+				FlagSet:   checkDocsiteFlagSet,
+				ShortHelp: "Check the code powering docs.sourcegraph.com for broken links and linting errors.",
+				Exec:      runCheckScriptsAndReport(scriptChecks["docsite"]...),
+			},
+			{
+				Name:      "docker",
+				FlagSet:   checkDockerFlagSet,
+				ShortHelp: "Check for forbidden docker base images",
+				Exec:      runCheckScriptsAndReport(scriptChecks["docker"]...),
+			},
+			{
+				Name:      "client",
+				FlagSet:   checkClientFlagSet,
+				ShortHelp: "Check client code for linting errors, forbidden imports, ...",
+				Exec:      runCheckScriptsAndReport(scriptChecks["client"]...),
+			},
+			{
+				Name:      "shell",
+				FlagSet:   checkShellFlagSet,
+				ShortHelp: "Check shell code for linting errors, formatting, ...",
+				Exec:      runCheckScriptsAndReport(scriptChecks["shell"]...),
+			},
+		},
+	}
+)
+
+type checkScriptFn func(context.Context) *checkReport
+
+// runCheckScriptsAndReport concurrently runs all fns and report as each check finishes. Returns an error
+// if any of the fns fails.
+func runCheckScriptsAndReport(fns ...checkScriptFn) func(context.Context, []string) error {
+	return func(ctx context.Context, _ []string) error {
+		_, err := root.RepositoryRoot()
+		if err != nil {
+			return err
+		}
+
+		// swawn a go routine for each check and increment count to report completion
+		var count int64
+		total := len(fns)
+		pending := out.Pending(output.Linef("", output.StylePending, "Running checks (done: 0/%d)", total))
+		var wg sync.WaitGroup
+		reportsCh := make(chan *checkReport)
+		wg.Add(total)
+		for _, fn := range fns {
+			go func(fn func(context.Context) *checkReport) {
+				report := fn(ctx)
+				atomic.AddInt64(&count, 1)
+				reportsCh <- report
+				wg.Done()
+			}(fn)
+		}
+		go func() {
+			wg.Wait()
+			close(reportsCh)
+		}()
+
+		// consume check reports
+		var hasErr bool
+		var messages []string
+		for report := range reportsCh {
+			pending.Destroy()
+			printCheckReport(report)
+			if count != int64(total) {
+				pending = out.Pending(output.Linef("", output.StylePending, "Running checks (done: %d/%d)", count, total))
+			}
+			if report.err != nil {
+				messages = append(messages, report.err.Error())
+				hasErr = true
+			}
+		}
+		// return the final error, if any
+		if hasErr {
+			return errors.Newf("failed checks: %s", strings.Join(messages, ", "))
+		}
+		return nil
+	}
+}
+
+type checkReport struct {
+	duration time.Duration
+	header   string
+	output   string
+	err      error
+}
+
+func printCheckReport(report *checkReport) {
+	timing := fmt.Sprintf(" (%ds)", report.duration/time.Second)
+	if report.err != nil {
+		writeFailureLine(report.header + timing)
+		out.Write(report.output)
+		return
+	}
+	writeSuccessLine(report.header + timing)
```

Comment:
```suggestion
	msg := fmt.Sprintf("%s (%ds)", report.header, report.duration/time.Second)
	if report.err != nil {
		writeFailureLine(msg)
		out.Write(report.output)
		return
	}
	writeSuccessLine(msg)
```

(btw. `writeFailureLine` etc are "printf"-style functions - I just noticed I should've added a `f` at their end :D)